tests/hazmat/primitives/test_pkcs12.py

```diff
@@ -290,43 +336,44 @@ class TestPKCS12Creation:
             (serialization.NoEncryption(), None),
         ],
     )
-    def test_generate(self, backend, name, encryption_algorithm, password):
-        cert, key = _load_ca(backend)
+    def test_generate_each_supported_keytype(
+        self, backend, ktype, kparam, name, encryption_algorithm, password
+    ):
+        if ktype.__module__.rsplit(".", 1)[-1].startswith("ed"):
```

Comment:
I don't think this is the right approach here at all. If you're looking to resist duplication then why not just make it parametrize over the callable that is the generation function and pass parameters if kparam is not None? Using `__module__` and `__import__` is definitely not something we want to land.

Reply:
doing it this way prevents the use of a third parameter, iiuc.  we pass the ktype both to this line and to the isinstance check on the key.  do i understand correctly that doing it the way you suggest would require parametrizing both the generator and the key type? can you help me understand your resistance to ```__module__``` and ```__import__``` in this context?

Reply:
The readability of this is lower than simply passing the callable and type. Readers need to go look up the module that a given ktype is a member of so they can see what the substring match does, as well as understand what `__import__` does and how it returns. And if that module name changes this would break in an unexpected way. That module name **won't** change, but we shouldn't depend on it this way.

Another argument for isinstance is fine :)

Reply:
this comment kind of addresses both points.  I wasnt able to do this before because I didn't notice the curve support check, but one of the reasons I like this code is because it paves the way to do nastily exhaustive things like this:
```
    @pytest.mark.parametrize(
        ("ktype", "kparam"),
        [
            pytest.param(
                ed448.Ed448PrivateKey,
                None,
                marks=pytest.mark.supported(
                    only_if=lambda backend: backend.ed448_supported(),
                    skip_message="Requires OpenSSL with Ed448 support",
                ),
            ),
            pytest.param(
                ed25519.Ed25519PrivateKey,
                None,
                marks=pytest.mark.supported(
                    only_if=lambda backend: backend.ed25519_supported(),
                    skip_message="Requires OpenSSL with Ed25519 support",
                ),
            ),
            (rsa.RSAPrivateKey, (65537, 1024)),
            (dsa.DSAPrivateKey, (1024,)),
        ]
        + [
            pytest.param(
                ec.EllipticCurvePrivateKey,
                (curve,),
                marks=pytest.mark.supported(
                    only_if=lambda backend: backend.elliptic_curve_supported(
                        curve
                    ),
                    skip_message=(
                        "Requires OpenSSL with " + curve.name + " support"
                    ),
                ),
            )
            for curve in [ec._CURVE_TYPES[i]() for i in ec._CURVE_TYPES]
        ],
    )
```